### PR TITLE
MM-31596: Support permalink in shared channels

### PR DESCRIPTION
**Pull request by agnivade**
We match permalinks in posts that are to be sent in shared channels.
If the permalink is not for the shared channel, we send an ephemeral
post to the post author notifying them that the link may not be
available to all remote users.
    
On the receiving side, we replace the site URL in the post
with the local site's URL.
    
```release-note
NONE
```

https://mattermost.atlassian.net/browse/MM-31596

### CODE REVIEW

File: services/sharedchannel/permalink.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			return msg
+		}
+
+		// Otherwise, modify pl to plshared as a marker to be replaced by remote sites
+		return strings.Replace(msg, "/pl/", "/"+permalinkMarker+"/", 1)
+	})
+}
+
+// processPermalinkFromRemote processes all permalinks coming from a remote site.
+func (scs *Service) processPermalinkFromRemote(msg string) string {
```
Comment (agnivade) [inline services/sharedchannel/permalink.go:55]: @wiggin77 - I assume I'll need to replace the remote team name with the local team name as well right?

File: services/sharedchannel/permalink.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+const (
+	permalinkMarker = "plshared"
+)
+
+// processPermalinkToRemote processes all permalinks going towards a remote site.
+func (scs *Service) processPermalinkToRemote(p *model.Post) string {
+	return permaLinkRegex.ReplaceAllStringFunc(p.Message, func(msg string) string {
+		// Extract the postID (This is simple enough not to warrant full-blown URL parsing.)
+		lastSlash := strings.LastIndexByte(msg, '/')
```
[the remark this answers is not in the document]
Comment (agnivade) [reply services/sharedchannel/permalink.go:30]: Yes, absolutely we can. And there is a test case for that too. Are you seeing any particular problem with this code?

File: services/sharedchannel/permalink.go (hunk 1 above)
Comment (wiggin77) [reply services/sharedchannel/permalink.go:60]: Correct.  RemoteTeamId is in the RemoteClusters table.

Comment (wiggin77) [reply services/sharedchannel/permalink.go:60]: Ah, looking closer I see this part is executed on the remote side.  Ignore my previous comment.  Do team names appear in perma-links?

Comment (wiggin77) [reply services/sharedchannel/permalink.go:60]: Ah they do. Yes, they would have to be replaced.

File: services/sharedchannel/permalink.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nronas) [reply services/sharedchannel/permalink.go:30]: i didn't see the test case in the first pass. But I can defo see it now, so all good

File: services/sharedchannel/permalink.go (hunk 1 above)
Comment (agnivade) [reply services/sharedchannel/permalink.go:60]: So to confirm, I should still take the team name from "RemoteTeamId is in the RemoteClusters table." ?

Comment (wiggin77) [reply services/sharedchannel/permalink.go:60]: Right, local team name for channel.

File: services/sharedchannel/permalink.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		}
+		if len(postList.Order) == 0 {
+			scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceWarn, "No post found for permalink", mlog.String("postID", postID))
+			return msg
+		}
+
+		// If postID is for a different channel
+		if postList.Posts[postList.Order[0]].ChannelId != p.ChannelId {
+			// Send ephemeral message to OP (only once per message).
+			// Note: Does this need to be translated?
```
Comment (wiggin77) [inline services/sharedchannel/permalink.go:45]: There is a ticket to have all ephemeral posts in the slash commands translated.  This should be as well, but I'm 0/5 whether that happens before or after MVP.  A ticket will suffice if that helps.

Comment (agnivade) [reply services/sharedchannel/permalink.go:45]: Did that in this ticket itself.

**Review by nronas — APPROVED** [review]
LGTM

**Review by wiggin77 — APPROVED** [review]
Looks great!  :+1:

**Pull request merged**